### PR TITLE
Add AWS SigV4 authentication support

### LINKED ISSUE #36: Proposal: AWS4 signatures signing
I need signed Opensearch requests to AWS, using AWS4 signatures. The Rust code to implement this is available (there are a couple of standalone crates, or we could extract it from the Rust AWS SDK). I would be willing to do the work, if you are willing to consider an PR. I'd welcome suggestions on how to implement, which signing code to use, how the API should work, etc.

### PR DESCRIPTION
**Pull request by Xtansia**
### Description
Implements authenticating requests via AWS SigV4 request signing.
 
I've taken a first attempt at this using the crate at https://github.com/awslabs/aws-sdk-rust/blob/main/sdk/aws-sigv4. I'm open to any feedback or improvements you might be able to offer on this. Main piece I'm unsure of would be the error handling, so if you can let me know what the desired manner would be, that'd be awesome.

### Issues Resolved
#36
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by rosscanning** [comment]
I originally requested this feature, and I can confirm that Xtansia's feature branch works for me.

**Comment by dblock** [comment]
@rosscanning Could you please amend your last commit with `-s` for DCO?

**Comment by dblock** [comment]
Let's document this, at least in README? And thanks!

**Comment by rosscanning** [comment]
@dblock I haven't done any commits. I just requested the feature; @Xtansia did the actual work :)

**Comment by dblock** [comment]
> @dblock I haven't done any commits. I just requested the feature; @Xtansia did the actual work :)

Yes, sorry I don't know what I was copy-pasting. Thanks @Xtansia !

Looks like we're seeing some test failures.

**Comment by Xtansia** [comment]
@dblock I've fixed that commit and added some further documentation. Will take a look into the test failures. I don't believe I've changed anything that would cause them to fail.

File: opensearch/src/http/aws_auth.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+};
+use reqwest::Request;
+
+const SERVICE_NAME: &str = "es";
+
+fn get_signing_params<'a>(
+    credentials: &'a Credentials,
+    region: &'a Region,
+) -> Result<SigningParams<'a>, BuildError> {
+    let mut settings = SigningSettings::default();
```
Comment (mpetri) [inline opensearch/src/http/aws_auth.rs:34]: could this be 

```
let settings = SigningSettings {
 percent_encoding_mode : PercentEncodingMode::Double;
 payload_checksum_kind : PayloadChecksumKind::NoHeader;
 signature_location : SignatureLocation::Headers;
 expires_in : None, 
 ..Default::default()
};

File: opensearch/src/http/aws_auth.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    let mut builder = SigningParams::builder()
+        .access_key(credentials.access_key_id())
+        .secret_key(credentials.secret_access_key())
+        .service_name(SERVICE_NAME)
+        .region(region.as_ref())
+        .time(SystemTime::now())
+        .settings(settings);
+
+    if let Some(session_token) = credentials.session_token() {
+        builder = builder.security_token(session_token);
```
Comment (mpetri) [inline opensearch/src/http/aws_auth.rs:49]: this could be replaced with `builder.set_security_token(credentials.session_token())` ?

Comment (mpetri) [reply opensearch/src/http/aws_auth.rs:49]: alternatively should this output an error message if this is `None`?

File: opensearch/src/http/aws_auth.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        &uri,
+        request.headers(),
+        SignableBody::Bytes(request.body().and_then(|b| b.as_bytes()).unwrap_or(&[])),
+    );
+
+    let (mut instructions, _) = sign(signable_request, &params)?.into_parts();
+
+    if let Some(new_headers) = instructions.take_headers() {
+        for (name, value) in new_headers.into_iter() {
+            request.headers_mut().insert(name.unwrap(), value);
```
Comment (mpetri) [inline opensearch/src/http/aws_auth.rs:77]: should this either be `expect` or return an error for the whole function?

File: opensearch/src/auth.rs [hunk 4]
```diff
@@ -45,6 +45,13 @@ pub enum Credentials {
     Certificate(ClientCertificate),
     /// An id and api_key to use for API key authentication
     ApiKey(String, String),
+    /// AWS credentials used for AWS SigV4 request signing.
+    /// 
+    /// # Optional
+    /// 
+    /// This requires the `aws-auth` feature to be enabled.
+    #[cfg(feature = "aws-auth")]
+    Aws(aws_types::credentials::SharedCredentialsProvider, aws_types::region::Region)
```
Comment (mpetri) [inline opensearch/src/auth.rs]: should this be `AwsSigV4` in case there is another Aws specific auth method?

**Review by mpetri — CHANGES_REQUESTED** [review]
Provided some general comments of some minor improvements I can see.

File: opensearch/src/http/aws_auth.rs (hunk 1 above)
Comment (Xtansia) [reply opensearch/src/http/aws_auth.rs:34]: `SigningSettings` is non_exhaustive, so unfortunately can not be constructed with a struct expression.

Comment (Xtansia) [reply opensearch/src/http/aws_auth.rs:34]: Though on closer inspection the values I'm setting are the defaults, so will simplify it.

File: opensearch/src/http/aws_auth.rs (hunk 2 above)
Comment (Xtansia) [reply opensearch/src/http/aws_auth.rs:49]: Yes, will replace it 👍 The `None` state of session_token is fine, as permanent credentials won't have one, while temporary ones such as EC2 instance credentials etc will have one.

File: opensearch/src/auth.rs (hunk 4 above)
Comment (Xtansia) [reply opensearch/src/auth.rs]: Happy to rename this. Should I rename the feature as well? Or keep a "generic" `aws-auth` feature and just make the `Credentials` variant more specific?

Comment (mpetri) [reply opensearch/src/auth.rs]: Hmm not sure what is better. If I was using this I would expect the `aws-auth` feature to enable all aws specific auth methods so maybe just making the Credentials variant more specific is fine?

File: opensearch/src/http/aws_auth.rs (hunk 3 above)
Comment (Xtansia) [reply opensearch/src/http/aws_auth.rs:77]: I can change this to `expect`. In practice this is never `None`, it's just due to the API of `HeaderMap`. This is exactly what the aws-sigv4 crate does when operating on an `http::Request` https://github.com/awslabs/aws-sdk-rust/blob/main/sdk/aws-sigv4/src/http_request/sign.rs#L135-L139

Comment (mpetri) [reply opensearch/src/http/aws_auth.rs:77]: yeah I understand it is never none. Maybe that is just my personal preference to never have `unwrap` in the code without a comment and then the comment might as well be in the `expect`

**Comment by Xtansia** [comment]
@dblock @mpetri I believe I've addressed all the comments so far. Additionally it appears the integration test failure was due to `simple_logger` or `time` crate versions, so I've updated the version used by `yaml_test_runner`. This failure was unrelated to my changes as it failed the check on this commit 2 months ago as well for the same reason https://github.com/opensearch-project/opensearch-rs/actions/runs/1844072847

**Comment by mpetri** [comment]
Looks okay to me!

File: opensearch/src/lib.rs [hunk 5]
```diff
@@ -58,6 +58,7 @@
 //! - **experimental-apis**: Enables experimental APIs. Experimental APIs are just that - an experiment. An experimental
```
Comment (dblock) [inline opensearch/src/lib.rs:58]: Is this some generated file that should be excluded from the repo?

**Review by dblock — APPROVED** [review]

File: opensearch/src/lib.rs (hunk 5 above)
Comment (dblock) [reply opensearch/src/lib.rs:58]: Nm, ignore me. Rust noob ;)

**Comment by dblock** [comment]
> Thank you @Xtansia for adding this feature!
> 
> Can we also add this usage in the [USER_GUIDE](https://github.com/opensearch-project/opensearch-rs/blob/main/USER_GUIDE.md)? It can be in another PR.

Good point. Let's do that as part of https://github.com/opensearch-project/opensearch-rs/issues/42.

**Pull request merged**